src/validation.cpp

```diff
@@ -5139,3 +5147,82 @@ class CMainCleanup
     }
 };
 static CMainCleanup instance_of_cmaincleanup;
+
+uint256 ChainstateManager::SnapshotBlockhash() const {
+    if (m_active_chainstate != nullptr) {
+        // If a snapshot chainstate exists, it will always be our active.
+        return m_active_chainstate->m_from_snapshot_blockhash;
+    }
+    return {};
+}
+
+std::vector<CChainState*> ChainstateManager::GetAll()
+{
+    std::vector<CChainState*> out;
+
+    if (!IsSnapshotValidated() && m_ibd_chainstate) {
+        out.push_back(m_ibd_chainstate.get());
+    }
+
+    if (m_snapshot_chainstate) {
+        out.push_back(m_snapshot_chainstate.get());
+    }
+
+    return out;
+}
+
+CChainState& ChainstateManager::InitializeChainstate(
+    bool activate, const uint256& snapshot_blockhash)
+{
+    std::unique_ptr<CChainState>& to_modify = (
+        snapshot_blockhash.IsNull() ? m_ibd_chainstate : m_snapshot_chainstate);
+
+    to_modify.reset(new CChainState(snapshot_blockhash));
+
+    if (activate) {
+        LogPrintf("Switching active chainstate to %s\n", to_modify->ToString());
+        m_active_chainstate = to_modify.get();
+    }
+
+    return *to_modify.get();
+}
+
+CChain& ChainstateManager::ActiveChain() const
+{
+    return m_active_chainstate->m_chain;
```

Comment:
in commit 16d20d91b43b1c30333fe14272d69cfb543229b3:

Should `assert m_active_chainstate` first?